Doxyigen documentation/man/man3/Player.3
.TH "Player" 3 "Tue Jan 10 2023" "Version 1" "Zuldesi" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Player
.SH SYNOPSIS
.br
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBPlayer\fP (WINDOW *win, int y, int x)"
.br
.ti -1c
.RI "int \fBgetmv\fP ()"
.br
.ti -1c
.RI "void \fBjump\fP ()"
.br
.ti -1c
.RI "int \fBairshoot\fP ()"
.br
.ti -1c
.RI "void \fBgodown\fP ()"
.br
.ti -1c
.RI "void \fBgoup\fP ()"
.br
.ti -1c
.RI "void \fBdisplay\fP ()"
.br
.ti -1c
.RI "void \fBinitialize\fP ()"
.br
.ti -1c
.RI "void \fBinjury\fP ()"
.br
.ti -1c
.RI "int \fBgetx\fP ()"
.br
.ti -1c
.RI "int \fBgety\fP ()"
.br
.ti -1c
.RI "int \fBgetCoins\fP ()"
.br
.ti -1c
.RI "int \fBgetDir\fP ()"
.br
.ti -1c
.RI "int \fBgetPoints\fP ()"
.br
.ti -1c
.RI "int \fBgetLife\fP ()"
.br
.ti -1c
.RI "bool \fBgetBuy\fP ()"
.br
.ti -1c
.RI "void \fBsetBuy\fP (bool val)"
.br
.ti -1c
.RI "void \fBSetJump\fP ()"
.br
.ti -1c
.RI "void \fBupdatePoints\fP (int \fBpoints\fP)"
.br
.ti -1c
.RI "void \fBupdateCash\fP (int \fBmoney\fP)"
.br
.ti -1c
.RI "void \fBupdateCoordinates\fP (int x, int y)"
.br
.ti -1c
.RI "void \fBsetDir\fP (int \fBdir\fP)"
.br
.ti -1c
.RI "\fBBullet\fP \fBgetBullet\fP ()"
.br
.ti -1c
.RI "\fBbullt\fP \fBsetBullet\fP (\fBbullt\fP tmp, int cod)"
.br
.ti -1c
.RI "\fBBullet\fP \fBgetExploBullet\fP ()"
.br
.ti -1c
.RI "\fBbullt\fP \fBsetExploBullet\fP (\fBbullt\fP tmp, int cod)"
.br
.ti -1c
.RI "\fBbullt\fP \fBshoot\fP (\fBbullt\fP tmp)"
.br
.ti -1c
.RI "\fBbullt\fP \fBexplo_shoot\fP (\fBbullt\fP tmp)"
.br
.ti -1c
.RI "\fBPowerup\fP \fBgetGun\fP ()"
.br
.ti -1c
.RI "\fBPowerup\fP \fBgetShield\fP ()"
.br
.ti -1c
.RI "\fBPowerup\fP \fBgetHP\fP ()"
.br
.ti -1c
.RI "\fBPowerup\fP \fBgetArmor\fP ()"
.br
.ti -1c
.RI "\fBPowerup\fP \fBgetTeleportation\fP ()"
.br
.ti -1c
.RI "\fBPowerup\fP \fBgetBullets\fP ()"
.br
.ti -1c
.RI "\fBPowerup\fP \fBgetJumping\fP ()"
.br
.ti -1c
.RI "\fBPowerup\fP \fBgetFly\fP ()"
.br
.ti -1c
.RI "\fBPowerup\fP \fBgetExplo_Bullets\fP ()"
.br
.ti -1c
.RI "void \fBsetGun\fP (string g)"
.br
.ti -1c
.RI "void \fBsetShield\fP (int s)"
.br
.ti -1c
.RI "void \fBsetHP\fP (int h)"
.br
.ti -1c
.RI "void \fBsetArmor\fP (int a)"
.br
.ti -1c
.RI "void \fBsetTeleportation\fP (int t)"
.br
.ti -1c
.RI "void \fBsetBullets\fP (int b)"
.br
.ti -1c
.RI "void \fBsetJumping\fP (int j)"
.br
.ti -1c
.RI "void \fBsetFly\fP (int f)"
.br
.ti -1c
.RI "void \fBsetExplo_Bullets\fP (int e)"
.br
.ti -1c
.RI "bool \fBgetActiveFly\fP ()"
.br
.ti -1c
.RI "void \fBsetActiveFly\fP (bool val)"
.br
.ti -1c
.RI "bool \fBgetActiveJump\fP ()"
.br
.ti -1c
.RI "int \fBgetTELEPORT_DISTANCE\fP (int i)"
.br
.ti -1c
.RI "int \fBgetFlyActiveDuration\fP ()"
.br
.ti -1c
.RI "void \fBsetFlyActiveDuration\fP (int val)"
.br
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "char \fBcharacter\fP [4] {}"
.br
.ti -1c
.RI "WINDOW * \fBcurwin\fP {}"
.br
.ti -1c
.RI "int \fBlife\fP {}"
.br
.ti -1c
.RI "int \fBcash\fP {}"
.br
.ti -1c
.RI "int \fBxLoc\fP {}"
.br
.ti -1c
.RI "int \fByLoc\fP {}"
.br
.ti -1c
.RI "int \fBxMax\fP {}"
.br
.ti -1c
.RI "int \fByMax\fP {}"
.br
.ti -1c
.RI "int \fBdir\fP {}"
.br
.ti -1c
.RI "int \fBpoints\fP {}"
.br
.ti -1c
.RI "int \fBjump_height\fP {}"
.br
.ti -1c
.RI "bool \fBactivejump\fP {}"
.br
.ti -1c
.RI "int \fBconta\fP {}"
.br
.ti -1c
.RI "int \fBsegno\fP {}"
.br
.ti -1c
.RI "bool \fBbuy\fP {}"
.br
.ti -1c
.RI "\fBPowerup\fP \fBgun\fP"
.br
.ti -1c
.RI "\fBPowerup\fP \fBshield\fP"
.br
.ti -1c
.RI "\fBPowerup\fP \fBhp\fP"
.br
.ti -1c
.RI "\fBPowerup\fP \fBarmor\fP"
.br
.ti -1c
.RI "\fBPowerup\fP \fBteleportation\fP"
.br
.ti -1c
.RI "\fBPowerup\fP \fBbullets\fP"
.br
.ti -1c
.RI "\fBPowerup\fP \fBjumping\fP"
.br
.ti -1c
.RI "\fBPowerup\fP \fBfly\fP"
.br
.ti -1c
.RI "\fBPowerup\fP \fBexplo_bullets\fP"
.br
.ti -1c
.RI "int \fBARMOR_DURATION\fP [3] {}"
.br
.ti -1c
.RI "int \fBTELEPORT_DISTANCE\fP [3] {}"
.br
.ti -1c
.RI "bool \fBACTIVE_ARMOR\fP {}"
.br
.ti -1c
.RI "int \fBARMOR_ACTIVE_DURATION\fP {}"
.br
.ti -1c
.RI "int \fBNUM_BULLETS\fP {}"
.br
.ti -1c
.RI "int \fBNUM_EXPLO_BULLETS\fP {}"
.br
.ti -1c
.RI "int \fBFLY_DURATION\fP {}"
.br
.ti -1c
.RI "int \fBFLY_ACTIVE_DURATION\fP {}"
.br
.ti -1c
.RI "bool \fBACTIVE_FLY\fP {}"
.br
.ti -1c
.RI "\fBBullet\fP \fBbullet\fP"
.br
.ti -1c
.RI "\fBBullet\fP \fBexplo_bullet\fP"
.br
.ti -1c
.RI "int \fBind\fP {}"
.br
.ti -1c
.RI "int \fBind2\fP {}"
.br
.in -1c
.SH "Member Function Documentation"
.PP 
.SS "int Player::airshoot ()"
Shoot during the jump movement or the down movement\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.SS "void Player::display ()"
See the player\&. 
.SS "\fBbullt\fP Player::explo_shoot (\fBbullt\fP tmp)"

.PP
\fBParameters\fP
.RS 4
\fItmp\fP 
.RE
.PP
\fBReturns\fP
.RS 4
.RE
.PP

.SS "bool Player::getActiveFly ()"
Take active fly variable\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.SS "bool Player::getActiveJump ()"
Take active jump variable\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.SS "\fBPowerup\fP Player::getArmor ()"

.PP
\fBReturns\fP
.RS 4
Armor\&. 
.RE
.PP

.SS "\fBBullet\fP Player::getBullet ()"
bullets\&. 
.PP
\fBReturns\fP
.RS 4
bullets\&. 
.RE
.PP

.SS "\fBPowerup\fP Player::getBullets ()"

.PP
\fBReturns\fP
.RS 4
Bullets\&. 
.RE
.PP

.SS "bool Player::getBuy ()"
Take buy variable\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.SS "int Player::getCoins ()"

.PP
\fBReturns\fP
.RS 4
coins of player\&. 
.RE
.PP

.SS "int Player::getDir ()"

.PP
\fBReturns\fP
.RS 4
dir of player\&. 
.RE
.PP

.SS "\fBPowerup\fP Player::getExplo_Bullets ()"

.PP
\fBReturns\fP
.RS 4
explosive bullets\&. 
.RE
.PP

.SS "\fBBullet\fP Player::getExploBullet ()"

.PP
\fBReturns\fP
.RS 4
explosive bullets\&. 
.RE
.PP

.SS "\fBPowerup\fP Player::getFly ()"

.PP
\fBReturns\fP
.RS 4
Fly\&. 
.RE
.PP

.SS "int Player::getFlyActiveDuration ()"
Take fly duration\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.SS "\fBPowerup\fP Player::getGun ()"
Get power-ups\&. 
.PP
\fBReturns\fP
.RS 4
Gun\&. 
.RE
.PP

.SS "\fBPowerup\fP Player::getHP ()"

.PP
\fBReturns\fP
.RS 4
HP\&. 
.RE
.PP

.SS "\fBPowerup\fP Player::getJumping ()"

.PP
\fBReturns\fP
.RS 4
Jump\&. 
.RE
.PP

.SS "int Player::getLife ()"
Life of player\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.SS "int Player::getmv ()"
Movement\&. 
.SS "int Player::getPoints ()"
Points of the player\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.SS "\fBPowerup\fP Player::getShield ()"

.PP
\fBReturns\fP
.RS 4
Shield\&. 
.RE
.PP

.SS "int Player::getTELEPORT_DISTANCE (int i)"
take the input parameter of the vector TELEPORT_DISTANCE 
.PP
\fBParameters\fP
.RS 4
\fIi\fP 
.RE
.PP
\fBReturns\fP
.RS 4
.RE
.PP

.SS "\fBPowerup\fP Player::getTeleportation ()"

.PP
\fBReturns\fP
.RS 4
Teleport\&. 
.RE
.PP

.SS "int Player::getx ()"
X coordinate of the player\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.SS "int Player::gety ()"
y coordinate of the player\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.SS "void Player::godown ()"
Fall\&. 
.SS "void Player::goup ()"
Go up\&. 
.SS "void Player::initialize ()"
Start the player\&. 
.SS "void Player::injury ()"
Damage\&. 
.SS "void Player::jump ()"
Jump\&. 
.SS "void Player::setActiveFly (bool val)"
Set active fly variable\&. 
.PP
\fBParameters\fP
.RS 4
\fIval\fP 
.RE
.PP

.SS "void Player::setArmor (int a)"
Set armor\&. 
.PP
\fBParameters\fP
.RS 4
\fIa\fP 
.RE
.PP

.SS "\fBbullt\fP Player::setBullet (\fBbullt\fP tmp, int cod)"
Set bullets\&. 
.PP
\fBParameters\fP
.RS 4
\fItmp\fP 
.br
\fIcod\fP 
.RE
.PP
\fBReturns\fP
.RS 4
.RE
.PP

.SS "void Player::setBullets (int b)"
Set bullets\&. 
.PP
\fBParameters\fP
.RS 4
\fIb\fP 
.RE
.PP

.SS "void Player::setBuy (bool val)"
Set buy variable\&. 
.PP
\fBParameters\fP
.RS 4
\fIval\fP 
.RE
.PP

.SS "void Player::setDir (int dir)"
Change direction\&. 
.PP
\fBParameters\fP
.RS 4
\fIdir\fP 
.RE
.PP

.SS "void Player::setExplo_Bullets (int e)"
Set explosive bullets\&. 
.PP
\fBParameters\fP
.RS 4
\fIe\fP 
.RE
.PP

.SS "\fBbullt\fP Player::setExploBullet (\fBbullt\fP tmp, int cod)"

.PP
\fBParameters\fP
.RS 4
\fItmp\fP 
.br
\fIcod\fP 
.RE
.PP
\fBReturns\fP
.RS 4
.RE
.PP

.SS "void Player::setFly (int f)"
Set fly\&. 
.PP
\fBParameters\fP
.RS 4
\fIf\fP 
.RE
.PP

.SS "void Player::setFlyActiveDuration (int val)"
Set fly duration\&. 
.PP
\fBParameters\fP
.RS 4
\fIval\fP 
.RE
.PP

.SS "void Player::setGun (string g)"
Set power-ups\&. Set Gun\&. 
.PP
\fBParameters\fP
.RS 4
\fIg\fP 
.RE
.PP

.SS "void Player::setHP (int h)"
Set HP\&. 
.PP
\fBParameters\fP
.RS 4
\fIh\fP 
.RE
.PP

.SS "void Player::SetJump ()"
Reset jump variable\&. 
.SS "void Player::setJumping (int j)"
Set jump\&. 
.PP
\fBParameters\fP
.RS 4
\fIj\fP 
.RE
.PP

.SS "void Player::setShield (int s)"
Set Shield\&. 
.PP
\fBParameters\fP
.RS 4
\fIs\fP 
.RE
.PP

.SS "void Player::setTeleportation (int t)"
Set teleport\&. 
.PP
\fBParameters\fP
.RS 4
\fIt\fP 
.RE
.PP

.SS "void Player::updateCash (int money)"
Update your wallet\&. 
.PP
\fBParameters\fP
.RS 4
\fImoney\fP 
.RE
.PP

.SS "void Player::updateCoordinates (int x, int y)"
Update your x and y when you went in a not-possible place\&. 
.PP
\fBParameters\fP
.RS 4
\fIx\fP 
.br
\fIy\fP 
.RE
.PP

.SS "void Player::updatePoints (int points)"
When you kill one enemy you take points\&. 
.PP
\fBParameters\fP
.RS 4
\fIpoints\fP 
.RE
.PP

.SH "Member Data Documentation"
.PP 
.SS "bool Player::ACTIVE_ARMOR {}\fC [protected]\fP"
Check if armor is activated\&. 
.SS "bool Player::ACTIVE_FLY {}\fC [protected]\fP"
Check if fly power-up is activated\&. 
.SS "bool Player::activejump {}\fC [protected]\fP"
If it is true, you are jumping\&. 
.SS "\fBPowerup\fP Player::armor\fC [protected]\fP"
Armor Power-up\&. 
.SS "int Player::ARMOR_ACTIVE_DURATION {}\fC [protected]\fP"
Armor time duration\&. 
.SS "int Player::ARMOR_DURATION[3] {}\fC [protected]\fP"
Constant for power-ups\&. Duration of the armor\&. 
.SS "\fBBullet\fP Player::bullet\fC [protected]\fP"
Gun magazine\&. Bullets magazine\&. 
.SS "\fBPowerup\fP Player::bullets\fC [protected]\fP"
Bullets power-up\&. 
.SS "bool Player::buy {}\fC [protected]\fP"
If you want to buy something\&. 
.SS "int Player::cash {}\fC [protected]\fP"
Money\&. 
.SS "char Player::character[4] {}\fC [protected]\fP"
\fBPlayer\fP's icon\&. 
.SS "int Player::dir {}\fC [protected]\fP"
Save the direction of the movement\&. 
.SS "\fBBullet\fP Player::explo_bullet\fC [protected]\fP"
Explosive bullets magazine\&. 
.SS "\fBPowerup\fP Player::explo_bullets\fC [protected]\fP"
Explosive bullets power-up\&. 
.SS "\fBPowerup\fP Player::fly\fC [protected]\fP"
Fly power-up\&. 
.SS "int Player::FLY_ACTIVE_DURATION {}\fC [protected]\fP"
Fly time duration remains\&. 

Player.cpp
/*
 * Player.cpp
 *
 *  Created on: 8 dic 2022
 *      Author: david
 */

#include "Player.hpp"


Player::Player(WINDOW * win, int y, int x){

	yLoc = y;
	xLoc = x;
	curwin = win;
	getmaxyx(curwin,yMax,xMax);
	keypad(curwin,true);
	character[0] = '@';
	character[1] = '^';
	character[2] = '|';
	character[3] = '-';
	life = 10;
	cash = 0;
	points = 0; //no points when you start
	dir = 1; //initial direction
	bullet = Bullet(); //initialize the bullet
	ind = 0; //no bullet
	ind2 = 0; //explosive bullet
	buy = false; //nothing to build when you start
	//jump
	activejump = false; //start without jump
	segno = -1; //segno for parabola
	conta = 1;
	jump_height = 5; //max height of the jump
	NUM_BULLETS = 0; //number of bullets
	NUM_EXPLO_BULLETS = 0; //number of exploding bullets
	//power-up
	gun = Powerup("None", "GUNS", 1, 10, 1); //no gun when you start
	shield = Powerup("Shield", "A shield that blocks damage one time", 0, 15, 1); //no shield when you start
	hp = Powerup("HP", "Get back on your feet one more time", life, 3, 1);
	armor = Powerup("Armor", "Become invincible for a limited time", 0, 10, 1); //no armor when you start
	teleportation = Powerup("Teleport", "Teleport a short distance", 0, 10, 1); //you cannot teleport when you start
	bullets = Powerup("Bullets","Charge of bullets",NUM_BULLETS,5,1); //no bullets when you start
	explo_bullets = Powerup("Explo-Bullets","Charge of explosive bullets",NUM_EXPLO_BULLETS,20,1); //no bullets when you start
	jumping = Powerup("Jump","Change the height of the jump",jump_height,7,1); //basic jump when you start
	fly = Powerup("Fly","You can fly",0,20,1); //you cannot fly when you start
	ARMOR_DURATION[0] = 100; ARMOR_DURATION[1] = 200; ARMOR_DURATION[2] = 500; //different type of Armor duration
	TELEPORT_DISTANCE[0] = dir*20; TELEPORT_DISTANCE[1] = dir*30; TELEPORT_DISTANCE[2] = dir*40; //different teleport distance
	ACTIVE_ARMOR = false; //no armor when you start
	ARMOR_ACTIVE_DURATION = 0; //no armor when you start
	FLY_ACTIVE_DURATION = 0; //no fly when you start
	FLY_DURATION = 500; //duration of the fly
	ACTIVE_FLY = false; //no fly when you start
}

Player::Player()= default; //default constructor

void Player::initialize(){
	display();
}

void Player::SetJump(){ //Rearrange jump variables
	conta = 1; //ready for the next jump
	segno = -1; //ready for the next jump
	activejump = false; //you reach the first floor
}

void Player::jump(){ //not move the x
	mvwaddch(curwin, yLoc, xLoc,' '); //Delete previous character
	if(armor.getQnt()>0) mvwaddch(curwin,yLoc-1,xLoc,' '); //Delete armor
	if(shield.getQnt()>0) mvwaddch(curwin,yLoc,xLoc-getDir(),' '); //delete shield
	if(shield.getQnt()==0) mvwaddch(curwin,yLoc,xLoc-getDir(),' '); //delete old gun
	if(gun.getName()!="none") mvwaddch(curwin,yLoc,xLoc+getDir(),' '); //delete gun
	if(conta<=2*(jumping.getQnt()-1)){
		if(conta == jumping.getQnt())segno = 1; //go down
		yLoc = yLoc + segno;
		conta++;
		if(yLoc<1){ //if you reach the roof
			SetJump(); //ready for a new jump
			while(yLoc<yMax-6)godown(); //godown
		}
	}
	else{
		SetJump(); //ready for a new jump
	}
}

void Player::godown(){ //go down
	mvwaddch(curwin, yLoc, xLoc,' '); //delete character
	if(armor.getQnt()>0) mvwaddch(curwin,yLoc-1,xLoc,' '); //Delete armor
	if(shield.getQnt()>0) mvwaddch(curwin,yLoc,xLoc-getDir(),' '); //delete shield
	if(shield.getQnt()==0) mvwaddch(curwin,yLoc,xLoc-getDir(),' '); //old gun
	if(strcmp(gun.getName().c_str(),"None")!=0) mvwaddch(curwin,yLoc,xLoc+getDir(),' '); //delete gun
	if(strcmp(gun.getName().c_str(),"Doublegun")==0) mvwaddch(curwin,yLoc,xLoc-getDir(),' '); //see gun
	yLoc = yLoc + 1; //you are falling
	if (yLoc>yMax-6) yLoc = yMax-6; //ground floor
}

void Player::goup(){ //go up
	mvwaddch(curwin, yLoc, xLoc,' '); //delete character
	if(armor.getQnt()>0) mvwaddch(curwin,yLoc-1,xLoc,' '); //Delete armor
	if(shield.getQnt()>0) mvwaddch(curwin,yLoc,xLoc-getDir(),' '); //delete shield
	if(shield.getQnt()==0) mvwaddch(curwin,yLoc,xLoc-getDir(),' '); //old gun
	if(strcmp(gun.getName().c_str(),"None")!=0) mvwaddch(curwin,yLoc,xLoc+getDir(),' '); //delete gun
	if(strcmp(gun.getName().c_str(),"Doublegun")==0) mvwaddch(curwin,yLoc,xLoc-getDir(),' '); //see gun
	yLoc = yLoc - 1; //you are lifting
	if(yLoc<1) yLoc = 1;
}

int Player::getmv(){ //move the character with gun by user
	if(ACTIVE_ARMOR){ //check if you have active your armor
		if(ARMOR_ACTIVE_DURATION>0)ARMOR_ACTIVE_DURATION--; //if you have the armor active, you have to decrement the time life of armor
		else ACTIVE_ARMOR = false; //Your armor finishes its life
	}
	int choice = wgetch(curwin);
	switch (choice){
		case KEY_UP: //just go up
			if(FLY_ACTIVE_DURATION>0){ //if you are flying
				goup();
			}
			else{ //if you are not flying
				activejump = true; //active jump
			}
			break;
		case KEY_DOWN: //just go down
			if(FLY_ACTIVE_DURATION>0){ //if you are flying
				godown();
			}
			break;
		case 'h': //activate the gun
			if(bullets.getQnt()>0 && strcmp(gun.getName().c_str(),"Pistol") == 0){ //you shoot one bullet
				bullet.blt = bullet.head_insert(bullet.blt,dir,xLoc,yLoc,ind); //add the bullet
				ind = ind + 1; //we want different indexes for the different bullets
				bullets.setQnt(bullets.getQnt()-1); //decrement bullets
			}
			else if(bullets.getQnt()>0 && strcmp(gun.getName().c_str(),"Rifle") == 0){ //you shoot two bullets
				for(int i=0;i<2;i++){
					bullet.blt = bullet.head_insert(bullet.blt,dir,xLoc+i*dir,yLoc,ind); //add the bullet
					ind = ind + 1; //we want different indexes for the different bullets
				}
				bullets.setQnt(bullets.getQnt()-1); //decrement bullets
			}
			else if(bullets.getQnt()>0 && strcmp(gun.getName().c_str(),"Machinegun") == 0){ //you shoot three bullets
				for(int i=0;i<3;i++){
					bullet.blt = bullet.head_insert(bullet.blt,dir,xLoc+i*dir,yLoc,ind); //add the bullet
					ind = ind + 1; //we want different indexes for the different bullets
				}
				bullets.setQnt(bullets.getQnt()-1); //decrement bullets
			}
			else if(bullets.getQnt()>0 && strcmp(gun.getName().c_str(), "Doublegun") == 0){//you shoot two bullets in opposite directions
				bullet.blt = bullet.head_insert(bullet.blt,dir,xLoc,yLoc,ind); //add the bullet
				ind = ind + 1; //we want different indexes for the different bullets
				bullets.setQnt(bullets.getQnt()-1); //decrement bullets
				bullet.blt = bullet.head_insert(bullet.blt,-dir,xLoc,yLoc,ind); //add the bullet
				ind = ind + 1;
			}
			break;
		case 'a': //activate the armor (time_life of armor starts)
			if(armor.getQnt()>0){ //you have at least on piece of armor
				ARMOR_ACTIVE_DURATION = ARMOR_DURATION[armor.getQnt()-1]; //thanks to the quantity you get the armor duration. For example if you have two pieces of armor, you have 1000 time duration
				armor.setQnt(0); //no armor anymore (you lose all pieces)
				ACTIVE_ARMOR = true;
			}
			break;
		case 'f': //activate the fly (time_life of fly starts)
			if(fly.getQnt()>0){ //you have the fly
				FLY_ACTIVE_DURATION = FLY_DURATION; //thanks to the quantity you get the fly duration.
				fly.setQnt(fly.getQnt()-1); //no fly anymore
				ACTIVE_FLY = true;
			}
			break;
		case 'e': //explosive bullets
			if(explo_bullets.getQnt()>0 && strcmp(gun.getName().c_str(),"None") != 0){ //you shoot one bullet
				explo_bullet.blt = bullet.head_insert(explo_bullet.blt,dir,xLoc,yLoc,ind2); //add the bullet
				ind2 = ind2 + 1; //we want different indexes for the different bullets
				explo_bullets.setQnt(explo_bullets.getQnt()-1); //decrement bullets
			}
			break;
		case 'b': //buy something in the market
			buy = true;
			break;
		default:
			break;
	}
	return choice;
}

int Player::airshoot(){ //during jump movement or down movement, you can just shoot
	int choice = wgetch(curwin);
	switch(choice){
		case 'h': //activate the gun
			if(bullets.getQnt()>0 && strcmp(gun.getName().c_str(),"Pistol") == 0){ //you shoot one bullet
				bullet.blt = bullet.head_insert(bullet.blt,dir,xLoc,yLoc,ind); //add the bullet
				ind = ind + 1; //we want different indexes for the different bullets
				bullets.setQnt(bullets.getQnt()-1); //decrement bullets
			}
			else if(bullets.getQnt()>0 && strcmp(gun.getName().c_str(),"Rifle") == 0){ //you shoot two bullets
				for(int i=0;i<2;i++){
					bullet.blt = bullet.head_insert(bullet.blt,dir,xLoc+i*dir,yLoc,ind); //add the bullet
					ind = ind + 1; //we want different indexes for the different bullets
				}
				bullets.setQnt(bullets.getQnt()-1); //decrement bullets
			}
			else if(bullets.getQnt()>0 && strcmp(gun.getName().c_str(),"Machinegun") == 0){ //you shoot three bullets
				for(int i=0;i<3;i++){
					bullet.blt = bullet.head_insert(bullet.blt,dir,xLoc+i*dir,yLoc,ind); //add the bullet
					ind = ind + 1; //we want different indexes for the different bullets
				}
				bullets.setQnt(bullets.getQnt()-1); //decrement bullets
			}
			else if(bullets.getQnt()>0 && strcmp(gun.getName().c_str(), "Doublegun") == 0){//you shoot two bullets in opposite directions
				bullet.blt = bullet.head_insert(bullet.blt,dir,xLoc,yLoc,ind); //add the bullet
				ind = ind + 1; //we want different indexes for the different bullets
				bullets.setQnt(bullets.getQnt()-1); //decrement bullets
				bullet.blt = bullet.head_insert(bullet.blt,-dir,xLoc,yLoc,ind); //add the bullet
				ind = ind + 1;
			}
			break;
		case 'a': //activate the armor (time_life of armor starts)
			if(armor.getQnt()>0){ //you have at least on piece of armor
				ARMOR_ACTIVE_DURATION = ARMOR_DURATION[armor.getQnt()-1]; //thanks to the quantity you get the armor duration. For example if you have two pieces of armor, you have 1000 time duration
				armor.setQnt(0); //no armor anymore (you lose all pieces)
				ACTIVE_ARMOR = true;
			}
			break;
		case 'e': //explosive bullets
			if(explo_bullets.getQnt()>0 && strcmp(gun.getName().c_str(),"None") != 0){ //you shoot one bullet
				explo_bullet.blt = bullet.head_insert(explo_bullet.blt,dir,xLoc,yLoc,ind2); //add the bullet
				ind2 = ind2 + 1; //we want different indexes for the different bullets
				explo_bullets.setQnt(explo_bullets.getQnt()-1); //decrement bullets
			}
			break;
		default:
			break;
	}
	return choice;
}

void Player::display(){ //display the character
	if(hp.getQnt()>1){
		wattron(curwin,COLOR_PAIR(3)); //color
		mvwaddch(curwin,yLoc,xLoc,character[0]); //see the player
		if(ACTIVE_ARMOR){
			mvwaddch(curwin,yLoc,xLoc,'O'); //see armor
		}
		if(shield.getQnt()>0) mvwaddch(curwin,yLoc,xLoc-getDir(),character[2]); //see the shield
		if(strcmp(gun.getName().c_str(),"None")!=0) mvwaddch(curwin,yLoc,xLoc+getDir(),character[3]); //see the gun
		if(strcmp(gun.getName().c_str(),"Doublegun")==0) mvwaddch(curwin,yLoc,xLoc-getDir(),character[3]); //see the gun
		wattroff(curwin,COLOR_PAIR(3)); //color
	}
	else{ //if you have one life, you become red
		wattron(curwin,COLOR_PAIR(1)); //color
		mvwaddch(curwin,yLoc,xLoc,character[0]); //see the player
		if(ACTIVE_ARMOR){
			mvwaddch(curwin,yLoc,xLoc,'O'); //see armor
		}
		if(shield.getQnt()>0) mvwaddch(curwin,yLoc,xLoc-getDir(),character[2]); //see the shield
		if(strcmp(gun.getName().c_str(),"None")!=0) mvwaddch(curwin,yLoc,xLoc+getDir(),character[3]); //see the gun
		if(strcmp(gun.getName().c_str(),"Doublegun")==0) mvwaddch(curwin,yLoc,xLoc-getDir(),character[3]); //see the gun
		wattroff(curwin,COLOR_PAIR(1));
	}
}


void Player::injury(){ //Injury
	if(shield.getQnt() == 0 && !ACTIVE_ARMOR){ //check if the player has the shield and the armor
		hp.setQnt(hp.getQnt()-1); // one point
	}
	else if(shield.getQnt() > 0 && !ACTIVE_ARMOR){ //check if the player has the shield but not the armor
		shield.setQnt(shield.getQnt()-1); //you lose one piece of shield
	}
	//if you have the armor, you are invincible
}

int Player::getx(){
	return xLoc;
}

int Player::gety(){
	return yLoc;
}

int Player::getLife(){
	return life;
}

int Player::getCoins() const{
	return cash;
}

int Player::getDir(){
	return dir;
}

int Player::getPoints() const{
	return points;
}

bool Player::getBuy(){
	return buy;
}

void Player::setBuy(bool val){
	this->buy = val;
}

void Player::updateCash(int money){ //update cash
	cash = cash + money;
}

void Player::updatePoints(int points){ 	//update money
	this->points = this->points + points;
}

void Player::updateCoordinates(int x,int y){ //update coordinates
	xLoc = xLoc+x;
	yLoc = yLoc+y;
}

void Player::setDir(int dir)
{
	this->dir=dir;
}

void Player::setLife(int life){
	this->life = life;
}

Bullet Player::getBullet(){
	return bullet;; //bullets
}

bullt Player::setBullet(bullt tmp,int cod){
	tmp = bullet.obj_remove(tmp,cod,false); //remove from the list (don't clean the memory)
	bullet.blt = bullet.obj_remove(bullet.blt,cod,true); //remove from the main list (clean the memory);
	return tmp;
}

Bullet Player::getExploBullet(){
	return explo_bullet; //explosive bullets
}

bullt Player::setExploBullet(bullt tmp,int cod){
	tmp = explo_bullet.obj_remove(tmp,cod,false); //remove from the list (don't clean the memory)
	explo_bullet.blt = explo_bullet.obj_remove(explo_bullet.blt,cod,true); //remove from the main list (clean the memory);
	return tmp;
}

bullt Player::shoot(bullt tmp){ //move bullets
	tmp = bullet.shoot(tmp,bullet.blt); //move bullets
	return tmp;
}

bullt Player::explo_shoot(bullt tmp){ //move explosive bullets
	tmp = bullet.shoot(tmp,explo_bullet.blt); //move explosive bullets
	return tmp;
}

bool Player::getActiveFly(){
	return ACTIVE_FLY; //take active fly variable
}
void Player::setActiveFly(bool val){
	ACTIVE_FLY = val; //set active fly variable
}
bool Player::getActiveJump(){
	return activejump; //take active jump variable
}
int Player::getTELEPORT_DISTANCE(int i){
	return TELEPORT_DISTANCE[i];
}
int Player::getFlyActiveDuration(){
	return FLY_ACTIVE_DURATION; //take fly duration
}
void Player::setFlyActiveDuration(int val){
	FLY_ACTIVE_DURATION = val; //set fly duration
}

//Power up

//get Power-ups
Powerup Player::getGun() const{
	return gun;
}
Powerup Player::getShield() const{
	return shield;
}
Powerup Player::getHP() const{
	return hp;
}
Powerup Player::getArmor() const{
	return armor;
}
Powerup Player::getTeleportation() const{
	return teleportation; //teleport
}
Powerup Player::getBullets() const{
	return bullets; //bullets
}
Powerup Player::getJumping() const{
	return jumping; //jump
}
Powerup Player::getFly() const{
	return fly; //fly
}
Powerup Player::getExplo_Bullets() const{
	return explo_bullets; //explosive bullets
}

//set Power-ups
void Player::setGun(string g){
	gun.setName(g); //gun
}

void Player::setShield(int s){
	shield.setQnt(s); //Shield
}
void Player::setHP(int h){
	hp.setQnt(h); //HP
}
void Player::setArmor(int a){
	armor.setQnt(a); //armor
}
void Player::setTeleportation(int t){
	teleportation.setQnt(t); //teleport
}
void Player::setBullets(int b){
	bullets.setQnt(b); //bullets
}
void Player::setJumping(int j){
	jumping.setQnt(j); //jump
}
void Player::setFly(int f){
	fly.setQnt(f); //fly
}
void Player::setExplo_Bullets(int e){
	explo_bullets.setQnt(e); //explosive bullets
}

